File: zeta/src/zeta/tools/bash.py
"""Bash tool implementation."""

import os
import signal
import subprocess
import time
from typing import Any

from zeta.capabilities.execution import error_result, proposed_command_effect
from zeta.capabilities.types import Capability, CapabilityId

DEFAULT_TIMEOUT_SECONDS = 120.0
MAX_TIMEOUT_SECONDS = 600.0
MAX_OUTPUT_CHARS = 12_000

SCHEMA: dict[str, Any] = {
    "type": "object",
    "additionalProperties": False,
    "required": ["command"],
    "properties": {
        "command": {"type": "string"},
        "reason": {"type": "string"},
        "timeout": {
            "type": "number",
            "minimum": 1,
            "maximum": MAX_TIMEOUT_SECONDS,
            "description": "Seconds before the command is killed (default 120).",
        },
    },
}

SPEC = Capability(
    CapabilityId("zeta", "bash"),
    "Execute or stage a shell command, depending on the active workflow.",
    SCHEMA,
)


def stage(params: dict[str, Any]) -> dict[str, Any]:
    command = str(params.get("command") or "").strip()
    if not command:
        return error_result("missing-command", "missing command")
    return proposed_command_effect(
        command,
        str(params.get("reason") or "Run the proposed command."),
    )


def run(params: dict[str, Any]) -> dict[str, Any]:
    command = str(params.get("command") or "").strip()
    if not command:
        return error_result("missing-command", "missing command")
    timeout_seconds = resolve_timeout(params.get("timeout"))
    started = time.monotonic()
    try:
        proc = subprocess.Popen(
            command,
            shell=True,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            start_new_session=True,
        )
    except OSError as exc:
        return error_result("bash-failed", str(exc))
    timed_out = False
    try:
        stdout_bytes, stderr_bytes = proc.communicate(timeout=timeout_seconds)
    except subprocess.TimeoutExpired:
        timed_out = True
        kill_process_group(proc)
        stdout_bytes, stderr_bytes = proc.communicate()
    duration_ms = int((time.monotonic() - started) * 1000)
    stdout, stdout_truncated = bounded_output(decode_output(stdout_bytes))
    stderr, stderr_truncated = bounded_output(decode_output(stderr_bytes))
    status = proc.returncode
    text = direct_output_text(
        command, status, stdout, stderr, timed_out=timed_out, timeout=timeout_seconds
    )
    result: dict[str, Any] = {
        "ok": status == 0 and not timed_out,
        "content": [
            {
                "type": "text",
                "text": text,
            }
        ],
        "metadata": {
            "mode": "direct",
            "command": command,
            "status": status,
            "duration_ms": duration_ms,
            "timed_out": timed_out,
            "stdout_truncated": stdout_truncated,
            "stderr_truncated": stderr_truncated,
        },
    }
    if timed_out:
        result["error"] = {
            "code": "bash-timeout",
            "message": (
                f"command timed out after {timeout_seconds:g}s and was killed"
            ),
        }
    elif result["ok"] is False:
        result["error"] = {
            "code": "bash-failed",
            "message": bash_failure_message(text, status),
        }
    return result


def resolve_timeout(value: Any) -> float:
    """Return the per-call timeout in seconds, clamped to the allowed range."""
    if isinstance(value, bool) or not isinstance(value, (int, float)):
        return DEFAULT_TIMEOUT_SECONDS
    return max(1.0, min(float(value), MAX_TIMEOUT_SECONDS))


def kill_process_group(proc: subprocess.Popen[bytes]) -> None:
    try:
        os.killpg(proc.pid, signal.SIGKILL)
    except (ProcessLookupError, PermissionError):
        proc.kill()


def decode_output(data: bytes | None) -> str:
    return (data or b"").decode("utf-8", errors="replace")


def bounded_output(text: str, limit: int = MAX_OUTPUT_CHARS) -> tuple[str, bool]:
    if len(text) <= limit:
        return text, False
    head = text[: limit // 2]
    tail = text[len(text) - limit // 2 :]
    omitted = len(text) - len(head) - len(tail)
    return f"{head}\n... {omitted} characters truncated ...\n{tail}", True


def direct_output_text(
    command: str,
    status: int,
    stdout: str,
    stderr: str,
    *,
    timed_out: bool = False,
    timeout: float = DEFAULT_TIMEOUT_SECONDS,
) -> str:
    sections = [
        f"$ {command}",
        f"exit {status}",
    ]
    if timed_out:
        sections.append(f"timed out after {timeout:g}s")
    if stdout:
        sections.extend(["stdout:", stdout])
    if stderr:
        sections.extend(["stderr:", stderr])
    return "\n".join(sections)


def bash_failure_message(text: str, status: int) -> str:
    return (
        bash_failure_summary(text) or flatten_tool_text(text) or f"exit status {status}"
    )


def bash_failure_summary(text: str) -> str:
    markers = (
        "error:",
        "Error:",
        "Exception:",
        "exceptions.",
        "TimeoutError:",
        "Unexpected",
        "No such file",
        "not found",
        "/bin/sh:",
    )
    for line in reversed(text.splitlines()):
        stripped = line.strip()
        if stripped.startswith("raise "):
            continue
        if any(marker in stripped for marker in markers):
            return stripped
    return ""


def flatten_tool_text(text: str) -> str:
    return " ".join(text.strip().split())

File: crates/zeta-agent/src/tools/bash.rs
//! Implements bounded shell execution with process-group deadlines.

use std::path::Path;
use std::time::Duration;

use serde_json::{json, Map, Value};

use super::{bounded_output, error_result, object, string_param, CommandOutput, CommandRunner};
use crate::{AbortSignal, AgentError};

const DEFAULT_TIMEOUT_SECONDS: f64 = 120.0;
const MAX_TIMEOUT_SECONDS: f64 = 600.0;
const MAX_OUTPUT_CHARS: usize = 12_000;

pub(super) fn bash<C: CommandRunner>(
    params: &Map<String, Value>,
    base_directory: &Path,
    commands: &mut C,
    abort: &dyn AbortSignal,
) -> Result<Map<String, Value>, AgentError> {
    let command = string_param(params, "command");
    let command = command.trim();
    if command.is_empty() {
        return Ok(error_result("missing-command", "missing command"));
    }
    let timeout = timeout_seconds(params.get("timeout"));
    let output = match commands.run_shell(
        command,
        base_directory,
        Duration::from_secs_f64(timeout),
        abort,
    ) {
        Ok(output) => output,
        Err(error) if error.kind() == std::io::ErrorKind::Interrupted => {
            return Err(AgentError::tool(error.to_string()))
        }
        Err(error) => return Ok(error_result("bash-failed", error.to_string())),
    };
    Ok(render_output(command, timeout, output))
}

fn timeout_seconds(value: Option<&Value>) -> f64 {
    let Some(value) = value.and_then(Value::as_f64) else {
        return DEFAULT_TIMEOUT_SECONDS;
    };
    value.clamp(1.0, MAX_TIMEOUT_SECONDS)
}

fn render_output(command: &str, timeout: f64, output: CommandOutput) -> Map<String, Value> {
    let (stdout, stdout_truncated) = bounded_output(&output.stdout, MAX_OUTPUT_CHARS);
    let (stderr, stderr_truncated) = bounded_output(&output.stderr, MAX_OUTPUT_CHARS);
    let text = direct_output_text(
        command,
        output.status,
        &stdout,
        &stderr,
        output.timed_out,
        timeout,
    );
    let ok = output.status == 0 && !output.timed_out;
    let mut result = object(json!({
        "ok": ok,
        "content": [{"type": "text", "text": text}],
        "metadata": {
            "command": command,
            "status": output.status,
            "duration_ms": output.duration_ms,
            "timed_out": output.timed_out,
            "stdout_truncated": stdout_truncated,
            "stderr_truncated": stderr_truncated,
        },
    }));
    if output.timed_out {
        result.insert(
            "error".to_owned(),
            json!({
                "code": "bash-timeout",
                "message": format!(
                    "command timed out after {}s and was killed",
                    display_seconds(timeout),
                ),
            }),
        );
    } else if !ok {
        result.insert(
            "error".to_owned(),
            json!({
                "code": "bash-failed",
                "message": failure_message(&text, output.status),
            }),
        );
    }
    result
}

fn direct_output_text(
    command: &str,
    status: i32,
    stdout: &str,
    stderr: &str,
    timed_out: bool,
    timeout: f64,
) -> String {
    let mut sections = vec![format!("$ {command}"), format!("exit {status}")];
    if timed_out {
        sections.push(format!("timed out after {}s", display_seconds(timeout)));
    }
    if !stdout.is_empty() {
        sections.push("stdout:".to_owned());
        sections.push(stdout.to_owned());
    }
    if !stderr.is_empty() {
        sections.push("stderr:".to_owned());
        sections.push(stderr.to_owned());
    }
    sections.join("\n")
}

fn failure_message(text: &str, status: i32) -> String {
    let summary = failure_summary(text);
    if !summary.is_empty() {
        return summary;
    }
    let flattened = flatten_text(text);
    if !flattened.is_empty() {
        return flattened;
    }
    format!("exit status {status}")
}

fn failure_summary(text: &str) -> String {
    let markers = [
        "error:",
        "Error:",
        "Exception:",
        "exceptions.",
        "TimeoutError:",
        "Unexpected",
        "No such file",
        "not found",
        "/bin/sh:",
    ];
    for line in text.lines().rev() {
        let line = line.trim();
        if line.starts_with("raise ") {
            continue;
        }
        for marker in markers {
            if line.contains(marker) {
                return line.to_owned();
            }
        }
    }
    String::new()
}

fn flatten_text(text: &str) -> String {
    let mut words = Vec::new();
    for word in text.split_whitespace() {
        words.push(word);
    }
    words.join(" ")
}

fn display_seconds(seconds: f64) -> String {
    if seconds.fract() == 0.0 {
        return format!("{seconds:.0}");
    }
    seconds.to_string()
}
